In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [5]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [11]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled_df = StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 
                                                                       'price_change_percentage_7d',
                                                                       'price_change_percentage_14d',
                                                                       'price_change_percentage_30d',
                                                                       'price_change_percentage_60d', 
                                                                       'price_change_percentage_200d',
                                                                       'price_change_percentage_1y']])

crypto_scaled_df [0:5]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.81092807,  0.24945797, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  2.00095907,  1.76061001,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

In [14]:
# Create a DataFrame with the scaled data
crypto_scaled_df = pd.DataFrame(crypto_scaled_df, columns=["price_change_percentage_24h", 
                    "price_change_percentage_7d","price_change_percentage_14d", "price_change_percentage_30d",
                    "price_change_percentage_60d", "price_change_percentage_200d", "price_change_percentage_1y"])


# Copy the crypto names from the original DataFrame
crypto_scaled_df['coin_id']=df_market_data.index

# Set the coin_id column as index
crypto_scaled_df = crypto_scaled_df.set_index("coin_id")

# Display the scaled DataFrame
crypto_scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [20]:
# Create a list with the number of k-values from 1 to 11
k= list(range(1,12))

k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [21]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
for i in k:
# 2. Fit the model to the data using `df_market_data_scaled`
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(crypto_scaled_df)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)

C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is k

In [23]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

elbow_df

k     inertia
0    1  287.000000
1    2  212.123342
2    3  165.136752
3    4   79.022435
4    5   66.413051
5    6   61.634088
6    7   46.074467
7    8   43.392694
8    9   32.913223
9   10   30.469111
10  11   26.480101

In [24]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_df.hvplot.line(
    x="k",
    y="inertia",
    title= "Elbow Curve",
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for k is k=4.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [64]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [65]:
# Fit the K-Means model using the scaled DataFrame
model.fit(crypto_scaled_df)

C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [66]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
k4 = model.predict(crypto_scaled_df)

# Print the resulting array of cluster values.
k4

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0])

In [67]:
# Create a copy of the scaled DataFrame
crypto_predicted_df=crypto_scaled_df.copy()

In [68]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
crypto_predicted_df['predicted_cluster'] = k4

# Display the copy of the scaled DataFrame
crypto_predicted_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  predicted_cluster  
coin_id                                                      
bitcoin                        -0.251637                  2  
ethereum                       -0.199352                  2  
tether                         -0.282061                  0  
ripple                         -0.295546                  0  
bitcoin-cash                   -0.270317                  2

In [74]:
# Create a scatter plot using hvPlot by setting
crypto_predicted_df.hvplot.scatter(
    
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
    x='price_change_percentage_24h',
    y='price_change_percentage_7d',
# Color the graph points with the labels found using K-Means and
    by='predicted_cluster',
# add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point.
    hover_cols = 'coin_id')

:NdOverlay   [predicted_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [75]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [76]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
crypto_pca = pca.fit_transform(crypto_predicted_df)

# View the scaled PCA data
crypto_pca[0:5]

array([[ 0.44890795, -1.24537573, -0.8506404 ],
       [ 0.49536709, -0.899823  , -1.31755943],
       [-0.81884571,  0.07189909,  0.69501507],
       [-0.84035711,  0.08005407,  0.54435952],
       [ 0.81324004, -2.66952035, -1.64332113]])

In [77]:
# Retrieve the explained variance to determine how much information can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.34871677, 0.31363391, 0.22627118])

In [78]:
sum(pca.explained_variance_ratio_)

0.8886218549859445

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8886218549859445


In [85]:
# Create a new DataFrame with the PCA data.
market_crypto_pca_df = pd.DataFrame(
    crypto_pca,
    columns = ["PCA1", "PCA2", "PCA3"]
)

# Copy the crypto names from the original scaled DataFrame
market_crypto_pca_df['coin_id'] = crypto_scaled_df.index

# Set the coin_id column as index
market_crypto_pca_df.set_index("coin_id", inplace=True)

# Display the scaled PCA DataFrame
market_crypto_pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin       0.448908 -1.245376 -0.850640
ethereum      0.495367 -0.899823 -1.317559
tether       -0.818846  0.071899  0.695015
ripple       -0.840357  0.080054  0.544360
bitcoin-cash  0.813240 -2.669520 -1.643321

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [87]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,12))

k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [88]:
# Create an empty list to store the inertia values
inertia_2 = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
for i in k:
# 2. Fit the model to the data using `df_market_data_pca`
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(market_crypto_pca_df)
# 3. Append the model.inertia_ to the inertia list
    inertia_2.append(k_model.inertia_)

C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is k

In [89]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_2 = {"k": k, "inertia": inertia_2}
# Create a DataFrame with the data to plot the Elbow curve
elbow_df_2 = pd.DataFrame(elbow_data_2)
elbow_df_2

k     inertia
0    1  290.622694
1    2  207.307439
2    3  123.126951
3    4   44.130602
4    5   33.217427
5    6   23.374713
6    7   20.943077
7    8   18.666961
8    9   13.640110
9   10   10.287149
10  11    7.958180

In [90]:
# Plot a line chart with all the inertia values computed with the different values of k to visually identify the optimal value for k.
elbow_df_2.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for k in this model is k=4.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it is the same k value as the original data. Although, it does have a smaller inertia value using PCA data than using the original data. This means that more clustered in the PCA data compared to the original data.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [91]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [92]:
# Fit the K-Means model using the PCA data
model.fit(market_crypto_pca_df)

C:\Users\yoian\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [93]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
crypto_predition_2 = model.predict(market_crypto_pca_df)

# Print the resulting array of cluster values.
crypto_predition_2

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 3, 1, 0, 0, 2, 0, 0, 0, 0])

In [94]:
# Create a copy of the scaled PCA DataFrame
crypto_predition_2_df= market_crypto_pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
crypto_predition_2_df['clusters_predition'] = crypto_predition_2

# Display the copy of the scaled PCA DataFrame
crypto_predition_2_df.head()

PCA1      PCA2      PCA3  clusters_predition
coin_id                                                       
bitcoin       0.448908 -1.245376 -0.850640                   1
ethereum      0.495367 -0.899823 -1.317559                   1
tether       -0.818846  0.071899  0.695015                   0
ripple       -0.840357  0.080054  0.544360                   0
bitcoin-cash  0.813240 -2.669520 -1.643321                   1

In [95]:
# Create a scatter plot using hvPlot by setting
crypto_predition_2_df.hvplot.scatter(
    
# `x="PC1"` and `y="PC2"`.
    x="PCA1",
    y="PCA2",
# Color the graph points with the labels found using K-Means and
    by = 'clusters_predition',
# add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point.
    hover_cols='coin_id'
)

:NdOverlay   [clusters_predition]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [98]:
# Composite plot to contrast the Elbow curves
(elbow_df.hvplot.line(x="k", y="inertia", title="Elbow Curve 1", xticks=k) + 
 elbow_df_2.hvplot.line(x="k", y="inertia", title="Elbow Curve 2", xticks=k))

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [100]:
# Composite plot to contrast the clusters
(crypto_predicted_df.hvplot.scatter(
    x='price_change_percentage_24h', 
    y='price_change_percentage_7d',  
    title="Cluster 1", 
    by='predicted_cluster', 
    hover_cols=['coin_id']) 
 + 
crypto_predition_2_df.hvplot.scatter(
    x="PCA1", 
    y="PCA2",
    title="Cluster 2", 
    by='clusters_predition', 
    hover_cols=['coin_id'])
)

:Layout
   .NdOverlay.I  :NdOverlay   [predicted_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [clusters_predition]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The elbow curve for the PCA-reduced data (Curve 2) shows a lower inertia compared to the curve for the original data (Curve 1). This suggests that clustering with fewer plots reduces the sum of squared distances within clusters, indicating improved clustering performance with tighter, more cohesive groupings that is easier to understand. As for the first Cluster graph that utilizes the original data that is scaled, the clusters appear overlapping, with no distinct separation between them. This suggests that the original dataset, with more data, may not have captured the data's underlying structure as easily. In contrast, the clusters in cluster graph 2 are well-separated and clearly defined. You can see four distinct clusters, suggesting that reducing the number of features through PCA has filtered out noise and revealed more meaningful patterns, leading to better clustering results. In conclusion, reducing the number of features using PCA has improved the clustering of the cryptocurrency data in K-Means. The clusters are more distinct and easier to separate, and the reduced feature set appears to have captured the essential data patterns while minimizing noise, resulting in more accurate clustering and readablilty.